### PR TITLE
 Customize equality in unit tests and better failure message

### LINKED ISSUE #2368: Better way to customize equality in unit tests and better failure message
Right now most test assertions delegate to Beam's `PAssert` and hamcrest. The most common feedback about writing tests are:

- Hard to apply custom comparison logic, e.g. comparing `List[T]` fields as sets
- Hard to parse failure message, e.g. `containsInAnyOrder` should contain _very long list of records_ but does not ...

For the 1st, cats `Eq[T]` seems like an idiomatic choice but we can't inject it into Beam/hamcrest easily. We either have to hack the internals of these java libs or add our own matchers?

2 is only doable if we add our own matchers instead of using Beam/hamcrest?

### PR DESCRIPTION
**Pull request by jto**
Fixes #2368 

### Pretty print common types (based on [lihaoyi/PPrint](https://github.com/lihaoyi/PPrint))
Example from `AutoCompleteTest`:

#### Before:
Screenshot 2020-03-04 at 11 27 35

#### After:
Screenshot 2020-03-04 at 11 17 44

#### Before:

Screenshot 2020-03-04 at 11 43 56

#### After:

Screenshot 2020-03-04 at 14 06 08

### Pretty print `SpecificRecord` instances

Example from `AvroExampleTest`:

#### Before:
Screenshot 2020-03-03 at 14 30 18
#### After:
Screenshot 2020-03-04 at 14 03 16

## Custom `Eq`

See `SCollectionMatchersTest`:

https://github.com/spotify/scio/blob/13b21c92b8a5f723c9146dbb96726b8dca35fc5e/scio-test/src/test/scala/com/spotify/scio/testing/SCollectionMatchersTest.scala#L592-L623

### CODE REVIEW

**Comment by jto** [comment]
Note: I did not touch `equalMapOf` because it seems that touching it would probably change its behaviour.

File: scio-test/src/main/scala/com/spotify/scio/testing/SCollectionMatchers.scala [hunk 1]
```diff
@@ -35,12 +35,65 @@ import org.hamcrest.MatcherAssert.assertThat
...
+    coll.map(t => TestWrapper(t))
+
+  implicit def testWrapperCoder[T: Coder: Eq]: Coder[TestWrapper[T]] =
+    Coder.gen[TestWrapper[T]]
+}
+
+private object ScioMatchers {
+  private def supplierFromCoder[A: Coder, B](@transient a: A)(builder: A => B) = {
+    val options = PipelineOptionsFactory.create().as(classOf[ScioOptions])
+    options.setNullableCoders(true)
```
Comment (regadas) [inline scio-test/src/main/scala/com/spotify/scio/testing/SCollectionMatchers.scala]: @jto can you add a comment on why do we need to set this?

Comment (jto) [reply scio-test/src/main/scala/com/spotify/scio/testing/SCollectionMatchers.scala]: Good catch! It was not necessary. Fixed it :)

**Comment by jto** [comment]
@nevillelyh wdyt ?

**Comment by nevillelyh** [comment]
How would it work in CI, i.e. if it doesn't support color output? `pprint` should auto detect and turn it off?

Also this is breaking, although IDK any lib using the API. Should be part of 0.9?

**Comment by jto** [comment]
@regadas Good point. I changed it

**Comment by jto** [comment]
@nevillelyh 

> How would it work in CI, i.e. if it doesn't support color output? pprint should auto detect and turn it off?

I would hope that any half decent CI would either ignore ANSI codes or actually support them. I'll check.

> Does this imply that users will have to import cats Eq instances for primitives etc? What about case classes? Using our magnolify-cats?

The default behaviour is to use java's `equals` (since `SCollectionMatchers extends EqInstances`). Any other implicit in scope can override the behaviour. 

> Also this is breaking, although IDK any lib using the API. Should be part of 0.9?

In Scio, it did not break anything. For the vast majority the upgrade should be source compatible. It is however indeed a breaking change (mostly bc. the return type of the matchers changed a bit).

**Comment by jto** [comment]
I think I can make it back compatible. Do not merge it yet.

**Comment by jto** [comment]
I don't like the fact that `TestWrapper` is visible in the API signature but making it a pure implementation details means that it has to be` Serialiazable` (cuz in a closure), which seems impossible.

**Comment by jto** [comment]
`TestWrapper` does not leak anymore 🎉

File: scio-test/src/main/scala/com/spotify/scio/testing/SCollectionMatchers.scala [hunk 2]
```diff
@@ -29,18 +29,149 @@ import org.apache.beam.sdk.transforms.SerializableFunction
...
+}
 
 /**
  * Trait with ScalaTest [[org.scalatest.matchers.Matcher Matcher]]s for
  * [[com.spotify.scio.values.SCollection SCollection]]s.
  */
-trait SCollectionMatchers {
+trait SCollectionMatchers extends EqInstances {
+
+  import ScioMatchers.{makeFn, makeFnSingle}
```
Comment (regadas) [inline scio-test/src/main/scala/com/spotify/scio/testing/SCollectionMatchers.scala]:
```suggestion
  import ScioMatchers.makeFn
```

Comment (regadas) [reply scio-test/src/main/scala/com/spotify/scio/testing/SCollectionMatchers.scala]: The other one is not used ... either this or replace `makeFn` with `ScioMatchers.makeFn` up to you 😄
